### PR TITLE
[NashMD] fix the edge case where the model is a peft model

### LINKED ISSUE #2781: NashMD trainer sampling policy wrong
### Reproduction

In `nash_md_trainer.py` Line 171, function `_generate_completions`:
```python
def _generate_completions(self, model, prompts):
      with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
          model_output = unwrapped_model.generate(
              input_ids=prompts["input_ids"],
              attention_mask=prompts["attention_mask"],
              generation_config=self.generation_config,
          )

          ref_model = model if self.ref_model is None else self.ref_model
          with torch.no_grad(), unwrap_model_for_generation(ref_model, self.accelerator) as unwrapped_ref_model:
              mixture_model = GeometricMixtureWrapper(
                  model=unwrapped_model,
                  ref_model=unwrapped_ref_model,
                  generation_config=self.generation_config,
                  mixture_coef=self.mixture_coef,
                  device=self.accelerator.device,
              )

              mixture_output = mixture_model.generate(
                  input_ids=prompts["input_ids"],
                  attention_mask=prompts["attention_mask"],
                  generation_config=self.generation_config,
              )

      return model_output, mixture_output
```
When we run the trainer with `model` being a PEFT model, and no `self.ref_model` provided (which means disabling the adapter gives the reference model), the `unwrapped_ref_model` in this function will be the model itself. A correct way is to pass `is_peft_model=<model is peft model>` in the call `unwrap_model_for_generation(ref_model, self.accelerator)`.

### System Info

- Platform: Linux-5.14.0-503.22.1.el9_5.x86_64-x86_64-with-glibc2.34
- Python version: 3.10.13
- PyTorch version: 2.2.1
- CUDA device(s): NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000
- Transformers version: 4.48.0
- Accelerate version: 1.2.1
- Accelerate config:
  - compute_environment: LOCAL_MACHINE
  - distributed_type: MULTI_GPU
  - mixed_precision: bf16
  - use_cpu: False
  - debug: False
  - num_processes: 8
  - machine_rank: 0
  - num_machines: 1
  - gpu_ids: all
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 3.2.0
- HF Hub version: 0.27.1
- TRL version: 0.13.0
- bitsandbytes version: not installed
- DeepSpeed version: 0.16.2
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: 0.0.2
- OpenAI version: not installed
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Fixes #2781 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by kashif** [comment]
@zhourunlong can you check if this fixes the edge case?

**Review by qgallouedec — APPROVED** [review]

**Review by lewtun — APPROVED** [review]
Thanks for the fix! LGTM

File: trl/trainer/nash_md_trainer.py [hunk 1]
```diff
@@ -170,28 +174,50 @@ def mixture_coef(self):
...
+        if self.ref_model is None:
+            # No explicit ref_model is provided.
+            # Use the base of the main `model` if it's a PEFT model.
+            # policy_model_for_gmw is already DDP-unwrapped.
+            if is_peft_available() and isinstance(policy_model_for_gmw, PeftModel):
+                ref_model_for_gmw = policy_model_for_gmw.get_base_model()
+            else:
+                # Not a PEFT model (or PEFT not available), or already a base model.
+                # Use the DDP-unwrapped policy model itself as the reference.
+                ref_model_for_gmw = policy_model_for_gmw
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/nash_md_trainer.py:201]: yes... i asked the orginal authors what to do in this case this was their suggestion...
